File: tests/run_browser.py
"""Run the browser regression with an isolated synthetic library and ephemeral port."""
import os
from pathlib import Path
import subprocess
import tempfile
import threading

from werkzeug.serving import WSGIRequestHandler, make_server
from test_server import create_app, fixture


class QuietHandler(WSGIRequestHandler):
    def log_request(self, *args, **kwargs):
        pass


def main():
    with tempfile.TemporaryDirectory(prefix='karaoke-browser-') as directory:
        root = Path(directory)
        fixture(root / 'Synthetic.mp4')
        app = create_app(root)
        server = make_server('127.0.0.1', 0, app, threaded=True, request_handler=QuietHandler)
        thread = threading.Thread(target=server.serve_forever, daemon=True)
        thread.start()
        try:
            env = {**os.environ, 'KARAOKE_TEST_URL': f'http://127.0.0.1:{server.server_port}'}
            subprocess.run(['node', str(Path(__file__).with_name('browser.cjs'))],
                           env=env, check=True, timeout=90)
        finally:
            server.shutdown()
            server.server_close()
            thread.join()
            app.extensions['karaoke_pool'].shutdown(wait=True)


if __name__ == '__main__':
    main()

File: tests/run-browser.ts
import { mkdtemp, rm } from 'node:fs/promises';
import path from 'node:path';
import os from 'node:os';
import { createApp, listen, serverUrl, closeServer } from '../src/server/app.js';
import { fixture, exec } from './fixtures.js';
const root = await mkdtemp(path.join(os.tmpdir(), 'karaoke-browser-'));
try {
  await fixture(path.join(root, 'Synthetic.mp4'));
  const { app, store } = await createApp(root, { webRoot: path.resolve('dist/web') });
  await store.scan();
  const server = await listen(app);
  try { const result = await exec(process.execPath, ['tests/browser.cjs'], { env: { ...process.env, KARAOKE_TEST_URL: serverUrl(server), KARAOKE_TEST_MEDIA_FILE: path.join(root, 'Synthetic.mp4') }, timeout: 90000 }); console.log(result.stdout); }
  finally { await closeServer(server); await store.close(); }
} finally { await rm(root, { recursive: true, force: true }); }
